include/aws/common/array_list.h

```diff
@@ -0,0 +1,265 @@
+#ifndef AWS_COMMON_ARRAY_LIST_H
+#define AWS_COMMON_ARRAY_LIST_H
+
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/common.h>
+#include <aws/common/error.h>
+#include <stdint.h>
+#include <string.h>
+
```

Comment:
Making this a big template thing seems overly complicated. Why not just stick the structure size into the array list structure, and return `void *`s?

Reply:
Because the whole semantic is to copy the data whatever it is. 

Casing an int to void * is cumbersome and error prone. Then we have to do memcpy everywhere and casts to uint8_t etc.... This is just safer if we are going to be doing value semantics.

Reply:
And this is a rare occasion, where it's getting used in lots of places, and almost always with value semantics.

Reply:
We already use pointer references in all get/put sites currently, looks like. It wouldn't make use sites any more cumbersome as far as I can see.